### PR TITLE
builtin: Add a new built-in functin: truncate

### PR DESCRIPTION
**Pull request by gaocegege**
Hi, I am a newcomer to TiDB :)

There was a PR #2886 about truncate, but it is unactive now, I think. Then I tries to  finish the work.

## Changes

1. Add a new parameter in MyDecimal.Round, to allow truncate in the same function.
2. Implement the eval function for truncate, which allows users to truncate decimals or float numbers.
3. Add test cases.

PTAL

Signed-off-by: Ce Gao <ce.gao@outlook.com>

<div align="center">
        Demo
</div>

### CODE REVIEW

**Comment by gaocegege** [comment]
And FYI, in my environment, make check will return an error:

```
bash gitcookie.sh
go get github.com/golang/lint/golint
vet
vet --shadow
./server/server_test.go:264: declaration of "err" shadows declaration at ./server/server_test.go:244
Makefile:69: recipe for target 'check' failed
make: *** [check] Error 1
```

My enviorment info:

```
➜  tidb git:(master) go version
go version go1.7.3 linux/amd64
➜  tidb git:(master) uname -a
Linux ist 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by coocood** [comment]
@gaocegege 
Thank you for your PR!

Define constants for 0, 5, 9 is better since it is used as an argument.

**Comment by gaocegege** [comment]
Yeah, it is. I'll update it.

**Comment by gaocegege** [comment]
PTAL

File: util/types/mydecimal.go [hunk 1]
```diff
@@ -41,6 +44,13 @@ const (
 
 	MaxFraction = 30
 	DivFracIncr = 4
+
+	// AlwaysRound rounds regardless of the decimal.
+	AlwaysRound RoundMode = 0
+	// NormalRound rounds normally.
+	NormalRound RoundMode = 5
+	// NeverRound just truncates the decimal.
+	NeverRound RoundMode = 9
```
Comment (coocood) [inline util/types/mydecimal.go]: According to 
https://github.com/mysql/mysql-server/blob/71f48ab393bce80a59e5a2e498cd1f46f6b43f9a/strings/decimal.c#L1638

The value 0 represents truncate, the value 5 represents half up.
We don't need `AlwaysRound`.

Comment (gaocegege) [inline util/types/mydecimal.go]: Sorry, does 0 represents truncate, not 9?

Comment (gaocegege) [inline util/types/mydecimal.go]: The code https://github.com/mysql/mysql-server/blob/71f48ab393bce80a59e5a2e498cd1f46f6b43f9a/strings/decimal.c#L1655 shows that 10 represents truncate, I'll take a look.

Comment (coocood) [inline util/types/mydecimal.go]: Well, truncate is 10 indeed.

Comment (coocood) [inline util/types/mydecimal.go]: 0 means increment, 5 means half up, any other values mean truncate.

**Comment by coocood** [comment]
@gaocegege 
We better not expose 0, 5, 10 in the argument, we can define HalfEven, Truncate as roundMode type.

**Comment by gaocegege** [comment]
Now I have defined three RoundMode, is that work?

> We better not expose 0, 5, 10 in the argument, we can define HalfEven, Truncate as roundMode type.

**Comment by coocood** [comment]
@gaocegege 
The 0, 5, 10 value is round digit, it's not round mode.
0 means the next digit >= 0, then increment,
5 means the next digit >=5 then increment.
10 means the next digit >= 10 then increment, as no single digits has value 10, it does not increment.
And depends on negative or positive, 0 and 10 can be swapped for FLOOR and CEILING mode.

**Comment by gaocegege** [comment]
Yeah， but IMO, mode is to hide the magic number from caller, and the implementation now works 🤔

**Comment by coocood** [comment]
@gaocegege 
The `Ceiling` round mode is not correct.

**Comment by coocood** [comment]
@gaocegege 
We better delete the `Ceiling` mode for now if we are not implementing ceiling function.

**Comment by gaocegege** [comment]
OK, I have removed Ceiling from Mode, and keep the implementation in Round function. Soft deletion is enough, IMO.

File: util/types/mydecimal.go [hunk 2]
```diff
@@ -759,7 +771,7 @@ func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
 		pos := wordsFracTo*digitsPerWord - frac - 1
 		shiftedNumber := to.wordBuf[toIdx] / powers10[pos]
 		digAfterScale := shiftedNumber % 10
-		if digAfterScale > roundDigit || (roundDigit == 5 && digAfterScale == 5) {
```
Comment (coocood) [inline util/types/mydecimal.go]: This operation `digAfterScale > int32(roundMode)` is unreasonable.
We should not use roundMode as a digit number.

File: util/types/mydecimal.go [hunk 3]
```diff
@@ -759,7 +771,7 @@ func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
...
+		if digAfterScale > int32(roundMode) || (roundMode == HalfEven && digAfterScale == 5) {
```
Comment (coocood) [inline util/types/mydecimal.go]: This operation digAfterScale > int32(roundMode) is unreasonable.
We should not use roundMode as a digit number.

**Comment by gaocegege** [comment]
Sorry for closing in accident :)

Now Mode is string, and roundDigit is set according to the value of roundMode as MySQL does.

PTAL

File: util/types/mydecimal.go [hunk 4]
```diff
@@ -690,6 +700,10 @@ func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
 	wordsInt := digitsToWords(int(d.digitsInt))
 
 	var roundDigit int32 = 5
+	/* TODO - fix this code as it won't work for CEILING mode */
```
Comment (coocood) [inline util/types/mydecimal.go]: To support CEILING mode in the future, switch round mode to set the roundDigit, then switch roundDigit at line 743 is the right way.

Comment (gaocegege) [inline util/types/mydecimal.go]: OK, I'll fix it.

**Comment by gaocegege** [comment]
PTAL

File: expression/builtin_math_test.go [hunk 5]
```diff
@@ -269,6 +269,43 @@ func (s *testEvaluatorSuite) TestRound(c *C) {
...
+		{[]interface{}{newDec("-1.23"), 0}, newDec("-1")},
+		{[]interface{}{newDec("-1.23"), 1}, newDec("-1.2")},
+		{[]interface{}{newDec("-11.23"), -1}, newDec("-10")},
+		{[]interface{}{newDec("1.58"), 0}, newDec("1")},
+		{[]interface{}{newDec("1.58"), 1}, newDec("1.5")},
+		{[]interface{}{newDec("11.58"), -1}, newDec("10")},
+		{[]interface{}{newDec("23.298"), -1}, newDec("20")},
+		{[]interface{}{newDec("23.298"), -100}, newDec("0")},
+		{[]interface{}{newDec("23.298"), 100}, newDec("23.298")},
+		{[]interface{}{nil, 2}, nil},
```
Comment (zhexuany) [inline expression/builtin_math_test.go]: add a overflow test case

File: expression/builtin_math.go [hunk 6]
```diff
@@ -1066,5 +1066,54 @@ type builtinTruncateSig struct {
...
+	}
+
+	// The number is a float, run math.Trunc.
+	x, err := args[0].ToFloat64(sc)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	// Get the truncate.
+	shift := math.Pow10(frac)
```
Comment (zhexuany) [inline expression/builtin_math.go]: Potential overflow here.

File: util/types/mydecimal.go [hunk 7]
```diff
@@ -41,6 +44,11 @@ const (
 
 	MaxFraction = 30
 	DivFracIncr = 4
+
+	// HalfEven rounds normally.
+	HalfEven RoundMode = "HalfEven"
+	// Truncate just truncates the decimal.
+	Truncate RoundMode = "Truncate"
```
Comment (zhexuany) [inline util/types/mydecimal.go]: add a Ceiling which can be used in switch statement somewhere below.

File: util/types/mydecimal.go [hunk 8]
```diff
@@ -726,8 +745,9 @@ func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
 	toIdx := wordsInt + wordsFracTo - 1
 	if frac == wordsFracTo*digitsPerWord {
 		doInc := false
-		switch roundDigit {
-		case 0:
+		switch roundMode {
+		// Notice: No support for ceiling mode now.
+		case "Ceiling":
```
Comment (zhexuany) [inline util/types/mydecimal.go]: This is the one should be replaced by `Ceiling`

**Comment by zhexuany** [comment]
@gaocegege Please rebase your br.

File: util/types/mydecimal.go [hunk 9]
```diff
@@ -726,8 +745,9 @@ func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
 	toIdx := wordsInt + wordsFracTo - 1
 	if frac == wordsFracTo*digitsPerWord {
 		doInc := false
-		switch roundDigit {
-		case 0:
+		switch roundMode {
```
Comment (coocood) [inline util/types/mydecimal.go]: Please swith `roundDigit` like:
```
switch roundDigit {
case 0:
case 5:
case 10:
}
```
The number 0, 5, 10 is used literally the digit number, do not need any constant representation.

File: util/types/mydecimal.go (hunk 7 above)
Comment (gaocegege) [inline util/types/mydecimal.go]: Since ceiling mode is not supported, I think it is better not to expose Ceiling as a mode here.

WDYT @coocood

Comment (coocood) [inline util/types/mydecimal.go]: If `Ceiling` is not exposed, then don't use it in `switch roundMode`.

Comment (gaocegege) [inline util/types/mydecimal.go]: Got it.

**Comment by gaocegege** [comment]
PTAL, I'm not sure if I understands the meaning of "overflow" @zhexuany

**Comment by zhexuany** [comment]
@gaocegege check this https://github.com/pingcap/tidb/blob/master/expression/builtin_math.go#L378

**Comment by gaocegege** [comment]
@zhexuany Then is https://github.com/pingcap/tidb/pull/3077/commits/cf6968a0880aca26b7a728b3e18464efe59e724d#diff-98616ef8b6b4a9e3d4a0f9e988fb9c78R53 work for you? 

I borrowed some code from https://github.com/pingcap/tidb/pull/3077/commits/cf6968a0880aca26b7a728b3e18464efe59e724d#diff-98616ef8b6b4a9e3d4a0f9e988fb9c78R40

File: util/types/mydecimal.go [hunk 10]
```diff
@@ -689,7 +699,16 @@ func (d *MyDecimal) Round(to *MyDecimal, frac int) (err error) {
 	wordsFrac := digitsToWords(int(d.digitsFrac))
 	wordsInt := digitsToWords(int(d.digitsInt))
 
-	var roundDigit int32 = 5
+	var roundDigit int32
+	/* TODO - fix this code as it won't work for CEILING mode */
+	switch roundMode {
+	case "Ceiling":
```
Comment (shenli) [inline util/types/mydecimal.go]: Could we define a const for Ceiling?

Comment (gaocegege) [inline util/types/mydecimal.go]: Yeah there is a mode `modeCeiling` in the newest commit.

**Comment by gaocegege** [comment]
PTAL

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM
@gaocegege Thanks!

**Pull request merged**